Gerbil-git fails to build on OpenBSD

The git checkout of gerbil fails to build on OpenBSD. The build crashes on crypto compilation:

```
... compile os/socket
... compile os/kqueue
... compile os/signal
... compile os/signal-handler
... compile os/pid
... compile net/bio/file
... compile net/bio
... compile xml/ssax
... compile xml
... compile crypto/libcrypto
libcrypto__0.c:3615:10: fatal error: 'libcrypto-rfc5114.c' file not found
#include "libcrypto-rfc5114.c"
         ^~~~~~~~~~~~~~~~~~~~~
1 error generated.
*** ERROR IN ##main -- target compilation or link failed while compiling "/home/tmy/git/gerbil/lib/std/crypto/libcrypto__0.scm"
*** ERROR IN #:loop38665, "misc/concurrent-plan.ss"@90.7 -- Failure to build (gxc: "crypto/libcrypto" (submodules: ((static-include: "crypto/libcrypto-rfc5114.c"))) "-cc-options" "" "-ld-options" " -lcrypto" "-e" "(include \"~~lib/_gambit#.scm\")") "Compilation error; gsc exit with nonzero status"
*** ERROR; build failed
tmy@asteroid gerbil $
```

Same issue is with 0.17 version.

Ah, the previous error was my mistake in gambit port Makefile, I changed the COMPILER but forgot to change CC parameter. I guess this resulted the compilation to use mix of system gcc (4.2.1) and ports-gcc (8.4.1). Fixed it and I compiled the gambit port with following diff:

```
diff 3c1d0e6f7c1a7453566310164ea13533318a7bd7 /usr/ports
blob - 38a5b472c77244c78da2a7604d10b69eb2d120f7
file + lang/gambit/Makefile
--- lang/gambit/Makefile
+++ lang/gambit/Makefile
@@ -3,7 +3,7 @@ BROKEN-riscv64= gsi and gsi-script segfault at startup
 COMMENT=       complete, efficient and reliable implementation of Scheme
 
 V=             4.9.4
-REVISION=      0
+REVISION=      1
 DISTNAME=      gambit-v${V:S/./_/g}
 PKGNAME=       gambit-$V
 API_V=         ${V:R:S/./0/}00${V:E}
@@ -23,7 +23,7 @@ EXTRACT_SUFX= .tgz
 
 WANTLIB +=     c crypto m ssl util
 
-COMPILER=      base-clang ports-gcc
+COMPILER=      ports-gcc
 COMPILER_LANGS=        c
 
 CONFIGURE_STYLE=gnu
@@ -34,14 +34,19 @@ CONFIGURE_ARGS+=--disable-cplusplus \
                --libdir=${PREFIX}/lib/gambit/ \
                --enable-c-opt="${CFLAGS}" \
                --enable-openssl \
-               --enable-poll
+               --enable-poll \
+               --enable-march='' \
+               --enable-dynamic-clib \
+               --enable-single-host
 
-CONFIGURE_ENV= CC=${CC} CXX=${CXX} CFLAGS="${CFLAGS}" CXXFLAGS="${CXXFLAGS}"
+CONFIGURE_ENV= CC=egcc CXX=eg++ CFLAGS="${CFLAGS}" CXXFLAGS="${CXXFLAGS}"
 
 WRKDIST=       ${WRKDIR}/gambit-$V
 
 MAKE_FILE=     ${WRKSRC}/makefile
 
+WRKDIST=       ${WRKDIR}/${PKGNAME}
+
 # On sparc64, gcc crashes with the default stack limit:
 # "cc: Internal error: Illegal instruction" on _gambc.c
 do-build:
```

With above the resulting gsc compiles code without spewing out the "unknown argument" error so I'm assuming it is working as expected.

Though with above the Gerbil compile still fails with "file not found" error when compiling libcrypto:

```
... compile net/bio/file
... compile net/bio
... compile xml/ssax
... compile xml
... compile crypto/libcrypto
libcrypto__0.c:3615:10: fatal error: libcrypto-rfc5114.c: No such file or directory
 #include "libcrypto-rfc5114.c"
          ^~~~~~~~~~~~~~~~~~~~~
compilation terminated.
*** ERROR IN ##main -- target compilation or link failed while compiling "/home/tmy/git/gerbil/lib/std/crypto/libcrypto__0.scm"
*** ERROR IN #:loop38665, "misc/concurrent-plan.ss"@90.7 -- Failure to build (gxc: "crypto/libcrypto" (submodules: ((static-include: "crypto/libcrypto-rfc5114.c"))) "-cc-options" "" "-ld-options" " -lcrypto" "-e" "(include \"~~lib/_gambit#.scm\")") "Compilation error; gsc exit with nonzero status"
*** ERROR; build failed
```

> Now if you can just please test the --with-openssl=/usr/local

On where should that be given? Gambit has '--enable-openssl' but it does not take any argument. Gerbil does not seem to have that either.

I'm not using GERBIL_BUILD_CORES on my builds.
I'm building gerbil in the src dir:

```
tmy@asteroid src $ CPPFLAGS=-I/usr/local/include LDFLAGS=-L/usr/local/lib GAMBCOMP_VERBOSE=1 CC=egcc ./build.sh stdlib  
[*] Building gerbil stdlib
... compile build-config
... compile xml
... compile crypto/libcrypto
libcrypto__0.c:3615:10: fatal error: libcrypto-rfc5114.c: No such file or directory
 #include "libcrypto-rfc5114.c"
          ^~~~~~~~~~~~~~~~~~~~~
compilation terminated.
*** ERROR IN ##main -- target compilation or link failed while compiling "/home/tmy/git/gerbil/lib/std/crypto/libcrypto__0.scm"
*** ERROR IN #:loop3833, "misc/concurrent-plan.ss"@90.7 -- Failure to build (gxc: "crypto/libcrypto" (submodules: ((static-include: "crypto/libcrypto-rfc5114.c"))) "-cc-options" "" "-ld-options" " -lcrypto" "-e" "(include \"~~lib/_gambit#.scm\")") "Compilation error; gsc exit with nonzero status"
*** ERROR; build failed
```

I've been building gerbil against ported gambit or custom compilation of gambit but I'm hitting the same "file not found" error on all cases.

I used the pretty much same build script as you, I changed paths and tweaked gerbil build as follows:
```
build-gerbil() {
    if [[ ! -d ${GERBIL_SRC} ]]; then
        git clone https://github.com/vyzo/gerbil ${GERBIL_SRC}
    fi
    #alias gcc=/usr/local/bin/egcc
    #ln -s /usr/local/bin/egcc /usr/local/bin/gcc||true
    cd ${GERBIL_SRC}/src

    git fetch -a && git clean -xfd && git reset --hard
    git checkout ${GERBIL_VERSION}
    sed -i 's/gcc/egcc/g' ${GERBIL_SRC}/src/build.sh
    ./configure \
        --prefix=${GERBIL_HOME} \
        --enable-libxml \
        --enable-libyaml
    ./build.sh
    ./install
}

```

With it I still get same error:

```
... compile xml/ssax
... compile xml
... compile crypto/libcrypto
libcrypto__0.c: In function 'ffi_create_HMAC_CTX':
libcrypto__0.c:3537:46: error: invalid application of 'sizeof' to incomplete type 'HMAC_CTX' {aka 'struct hmac_ctx_st'}
   HMAC_CTX *ctx = (HMAC_CTX*)malloc (sizeof (HMAC_CTX));
                                              ^~~~~~~~
libcrypto__0.c:3539:5: warning: implicit declaration of function 'HMAC_CTX_init'; did you mean 'HMAC_CTX_new'? [-Wimplicit-function-declaration]
     HMAC_CTX_init (ctx);
     ^~~~~~~~~~~~~
     HMAC_CTX_new
libcrypto__0.c: In function 'ffi_release_HMAC_CTX':
libcrypto__0.c:3550:3: warning: implicit declaration of function 'HMAC_CTX_cleanup'; did you mean 'HMAC_CTX_get_md'? [-Wimplicit-function-declaration]
   HMAC_CTX_cleanup ((HMAC_CTX*) ptr);
   ^~~~~~~~~~~~~~~~
   HMAC_CTX_get_md
libcrypto__0.c: In function 'ffi_DH_pub_key':
libcrypto__0.c:3601:11: error: dereferencing pointer to incomplete type 'DH' {aka 'struct dh_st'}
  return dh->pub_key;
           ^~
libcrypto__0.c: At top level:
libcrypto__0.c:3615:10: fatal error: libcrypto-rfc5114.c: No such file or directory
 #include "libcrypto-rfc5114.c"
          ^~~~~~~~~~~~~~~~~~~~~
compilation terminated.
*** ERROR IN ##main -- target compilation or link failed while compiling "/home/tmy/tmp/gerbil-src/lib/std/crypto/libcrypto__0.scm"
*** ERROR IN #:loop38665, "misc/concurrent-plan.ss"@90.7 -- Failure to build (gxc: "crypto/libcrypto" (submodules: ((static-include: "crypto/libcrypto-rfc5114.c"))) "-cc-options" "" "-ld-options" " -lcrypto" "-e" "(include \"~~lib/_gambit#.scm\")") "Compilation error; gsc exit with nonzero status"
*** ERROR; build failed
 $
```

Though I'd guess the HMAC errors are due to changes in libressl which shouldn't be there if the build was using openssl instead of libressl.

--- a/src/gerbil/prelude/gambit/threads.ss
+++ b/src/gerbil/prelude/gambit/threads.ss
@@ -228,8 +228,10 @@ package: gerbil/gambit
   (let (handler (current-exception-handler))
     (with-exception-handler
      (lambda (e)
-       (mutex-unlock! mx)
-       (handler e)
+       (with-catch void
+         (lambda ()
+           (mutex-unlock! mx)
+           (handler e)))
        ;; if the handler returns here the state is inconsistent -- we need to bail
        (##thread-end-with-uncaught-exception! e))
      (lambda ()

--- a/src/gerbil/prelude/gambit/threads.ss
+++ b/src/gerbil/prelude/gambit/threads.ss
@@ -225,6 +225,20 @@ package: gerbil/gambit
   (thread-thread-group (current-thread)))
 
 (def (with-lock mx proc)
+  (let (handler (current-exception-handler))
+    (with-exception-handler
+     (lambda (e)
+       (mutex-unlock! mx)
+       (handler e)
+       ;; if the handler returns here the state is inconsistent -- we need to bail
+       (##thread-end-with-uncaught-exception! e))
+     (lambda ()
+       (mutex-lock! mx)
+       (let (result (proc))
+         (mutex-unlock! mx)
+         result)))))
+
+(def (with-dynamic-lock mx proc)
   (dynamic-wind
       (cut mutex-lock! mx)
       proc

--- a/src/std/build-std.ss
+++ b/src/std/build-std.ss
@@ -3,6 +3,8 @@
 ;;; Inner build file for std, as specially imported and called by ./build.ss for bootstrap reasons
 
 (import ./build-config ./make)
+;; use this when making changes to make and want to rebuild
+;; (import "build-config" "make")
 
 (include "build-spec.ss")
 

--- a/src/std/event.ss
+++ b/src/std/event.ss
@@ -140,7 +140,7 @@
       (mutex-unlock! mx))))
 
 (def (selector-wait-io-condvar sel timeo)
-  (##wait-for-io! sel (or timeo #t)))
+  (##wait-for-io! sel (if timeo (timeout->abs-timeout->seconds timeo) #t)))
 
 (def (maybe-timeout? obj)
   (or (not obj)
@@ -150,6 +150,11 @@
 (def (timeout->abs-timeout timeo)
   (if (time? timeo) timeo
       (seconds->time (+ (##current-time-point) timeo))))
+
+(def (timeout->abs-timeout->seconds timeo)
+  (if (time? timeo)
+    (time->seconds timeo)
+    (+ (##current-time-point) timeo)))
 
 (def (io-condition-variable? obj)
   (and (condition-variable? obj)

--- a/src/std/event.ss
+++ b/src/std/event.ss
@@ -140,7 +140,7 @@
       (mutex-unlock! mx))))
 
 (def (selector-wait-io-condvar sel timeo)
-  (##wait-for-io! sel (if timeo (time->seconds (timeout->abs-timeout timeo)) #t)))
+  (##wait-for-io! sel (or timeo #t)))
 
 (def (maybe-timeout? obj)
   (or (not obj)

--- a/src/std/interface.ss
+++ b/src/std/interface.ss
@@ -20,8 +20,8 @@
 (def +interface-prototypes+ (make-hash-table))
 (def +interface-prototypes-mx+ (make-mutex 'interface-constructor))
 
-;; create a new instance of an interface for an object
-(def (new descriptor obj)
+;; cast an object to an interface instance
+(def (cast descriptor obj)
   (declare (not safe))
   (let (klass (interface-descriptor-type descriptor))
     (cond
@@ -29,8 +29,8 @@
       ;; already an instance of the right interface
       obj)
      ((interface-instance? obj)
-      ;; another interface instance, cast
-      (new descriptor (interface-instance-object obj)))
+      ;; another interface instance, recast
+      (cast descriptor (interface-instance-object obj)))
      (else
       ;; vanilla object, convert to an interface instance
       (let* ((prototype-key (cons (##type-id klass) (type-of obj)))
@@ -48,7 +48,7 @@
                                  (or (method-ref obj method)
                                      (begin
                                        (mutex-unlock! +interface-prototypes-mx+)
-                                       (error "Cannot create interface prototype; missing method" (##type-name klass) method))))
+                                       (error "Cannot create interface instance; missing method" (##type-name klass) method))))
                                (interface-descriptor-methods descriptor)))
                          (prototype
                           (apply ##structure klass #f method-impls)))
@@ -77,12 +77,12 @@
                                   method-impl unchecked-method-impl))
 
   (defmethod {apply-macro-expander interface-info}
-    (with-syntax ((new (quote-syntax new)))
+    (with-syntax ((cast (quote-syntax cast)))
       (lambda (self stx)
         (syntax-case stx ()
           ((_ obj)
            (with-syntax ((descriptor (interface-info-descriptor self)))
-             #'(new descriptor obj))))))))
+             #'(cast descriptor obj))))))))
 
 (defsyntax (interface stx)
   (def symbol<?
@@ -268,11 +268,10 @@
                                               ((out ...) (method-arguments signature)))
                                   #'(begin
                                       (def (method self in ...)
-                                        (unless (predicate self)
-                                          (error "object is not an interface instance" self 'name))
-                                        (let ((obj (##unchecked-structure-ref self 1 klass 'method))
-                                              (f   (##unchecked-structure-ref self offset klass 'method)))
-                                          (f obj out ...)))
+                                        (let (instance (cast descriptor self))
+                                          (let ((obj (##unchecked-structure-ref instance 1 klass 'method))
+                                                (f   (##unchecked-structure-ref instance offset klass 'method)))
+                                            (f obj out ...))))
                                       (def (unchecked-method self in ...)
                                         (declare (not safe))
                                         (let ((obj (##unchecked-structure-ref self 1 klass 'method))
@@ -283,11 +282,10 @@
                                               ((out ...) (method-arguments signature)))
                                   #'(begin
                                       (def (method self . in)
-                                        (unless (predicate self)
-                                          (error "object is not an interface instance" self 'name))
-                                        (let ((obj (##unchecked-structure-ref self 1 klass 'method))
-                                              (f   (##unchecked-structure-ref self offset klass 'method)))
-                                          (apply f obj out ...)))
+                                        (let (instance (cast descriptor self))
+                                          (let ((obj (##unchecked-structure-ref instance 1 klass 'method))
+                                                (f   (##unchecked-structure-ref instance offset klass 'method)))
+                                            (apply f obj out ...))))
                                       (def (unchecked-method self . in)
                                         (declare (not safe))
                                         (let ((obj (##unchecked-structure-ref self 1 klass 'method))
@@ -313,7 +311,7 @@
                           (make-interface-descriptor klass '(method-name ...))))
                     (defmake
                       #'(def (make obj)
-                          (new descriptor obj)))
+                          (cast descriptor obj)))
                     (defpred
                       #'(def (predicate obj)
                           (direct-instance? klass obj)))

--- a/src/std/make.ss
+++ b/src/std/make.ss
@@ -612,6 +612,7 @@ TODO:
     ([static-exe: modf . opts]
      (compile-static-exe modf opts settings))
     ([static-include: file]
+     (copy-target file settings)
      (copy-static file settings))
     ([copy: file]
      (copy-compiled file settings))
@@ -778,6 +779,17 @@ TODO:
   (when (file-exists? spath)
     (delete-file spath))
   (copy-file file spath))
+
+(def (copy-target file settings)
+  (let* ((libdir (settings-libdir-prefix settings))
+         (srcdir (settings-srcdir settings))
+         (spath (path-expand file srcdir))
+         (tpath (path-expand file libdir)))
+    (message "... copy static include to target directory " file)
+    (create-directory* (path-directory tpath))
+    (when (file-exists? tpath)
+      (delete-file tpath))
+    (copy-file spath tpath)))
 
 (def (copy-compiled file settings)
   (def srcpath (source-path file #f settings))

--- a/src/std/make.ss
+++ b/src/std/make.ss
@@ -101,7 +101,7 @@ TODO:
       prefix: (prefix_ #f) force: (force? #f)
       optimize: (optimize #t) debug: (debug 'env)
       static: (static #t) static-debug: (static-debug #f)
-      verbose: (verbose #f) build-deps: (build-deps_ #f)
+      verbose: (verbose_ #f) build-deps: (build-deps_ #f)
       parallelize: (parallelize_ #f))
     (def gerbil-path (getenv "GERBIL_PATH" "~/.gerbil"))
     (def srcdir (or srcdir_ (error "srcdir must be specified")))
@@ -111,6 +111,11 @@ TODO:
     (def libdir-prefix (if prefix (path-expand prefix libdir) libdir))
     (def build-deps (path-expand (or build-deps_ "build-deps") srcdir))
     (def parallelize (gerbil-build-cores parallelize))
+    (def verbose
+      (cond
+       (verbose_)
+       ((getenv "GERBIL_BUILD_VERBOSE" #f) => string->number)
+       (else #f)))
     (struct-instance-init!
       self
       srcdir libdir bindir prefix force? optimize debug static static-debug verbose build-deps
@@ -592,6 +597,9 @@ TODO:
   (match spec
     ((? string? modf)
      (gxc-compile modf #f settings #t))
+    ([gxc: modf [submodules: submodules . _] . opts]
+     (for-each (cut build <> settings) submodules)
+     (gxc-compile modf opts settings #t))
     ([gxc: modf . opts]
      (gxc-compile modf opts settings #t))
     ([gsc: modf . opts]

--- a/src/std/net/httpd-test.ss
+++ b/src/std/net/httpd-test.ss
@@ -13,7 +13,7 @@
   (test-suite "test :std/net/httpd"
 
     (def server-address
-      "127.0.0.1:9999")
+      "127.0.0.1:19999")
 
     (def server-url
       (string-append "http://" server-address))

--- a/src/std/os/socket.ss
+++ b/src/std/os/socket.ss
@@ -92,6 +92,34 @@
         SO_TIMESTAMP
         SO_USELOOPBACK
 
+        IPPROTO_IP
+        IP_ADD_MEMBERSHIP
+        IP_DROP_MEMBERSHIP
+        IP_ADD_SOURCE_MEMBERSHIP
+        IP_DROP_SOURCE_MEMBERSHIP
+        IP_BLOCK_SOURCE
+        IP_UNBLOCK_SOURCE
+        IP_FREEBIND
+        IPPROTO_IP IP_HDRINCL
+        IP_MTU
+        IP_MTU_DISCOVER
+        IP_MULTICAST_ALL
+        IP_MULTICAST_IF
+        IP_MULTICAST_LOOP
+        IP_MULTICAST_TTL
+        IP_NODEFRAG
+        IP_OPTIONS
+        IP_PKTINFO
+        IP_RECVERR
+        IP_RECVORIGDSTADDR
+        IP_RECVOPTS
+        IP_RECVTOS
+        IP_RECVTTL
+        IP_RETOPTS
+        IP_ROUTER_ALERT
+        IP_TOS
+        IP_TTL
+
         IPPROTO_IPV6
         IPV6_ADDRFORM
         IPV6_ADD_MEMBERSHIP
@@ -365,7 +393,7 @@
       (error "Unknown address family" sa saf)))))
 
 ;;; sockopts
-(def (socket-getsockopt sock level opt . args)
+(def (socket-getsockopt sock level opt)
   (cond
    ((hash-get socket-sockopts level)
     => (lambda (ht)
@@ -374,14 +402,14 @@
            => (match <>
                 ((values getf setf)
                  (if getf
-                   (apply getf sock level opt args)
+                   (getf sock level opt)
                    (error "No getsockopt operation defined for option" level opt)))))
           (else
            (error "Unknown socket option" level opt)))))
    (else
     (error "Unknown socket level" level opt))))
 
-(def (socket-setsockopt sock level opt val . args)
+(def (socket-setsockopt sock level opt val)
   (cond
    ((hash-get socket-sockopts level)
     => (lambda (ht)
@@ -390,7 +418,7 @@
            => (match <>
                 ((values getf setf)
                  (if setf
-                   (apply setf sock level opt val args)
+                   (setf sock level opt val)
                    (error "No setsockopt operation defined for option" level opt)))))
           (else
            (error "Unknown socket option" level opt)))))
@@ -479,8 +507,8 @@
   (match ips
     ((cons maddr ifindex)
      (let (maddr (ip6-address maddr))
-     (check-os-error (_setsockopt_mreq6 (fd-e sock) level opt maddr ifindex)
-       (socket-setsockopt sock level opt ips))))
+       (check-os-error (_setsockopt_mreq6 (fd-e sock) level opt maddr ifindex)
+         (socket-setsockopt sock level opt ips))))
     (else
      (error "Bad argument; expected pair of ip6 addresses" ips))))
 

